docs/auth.md

```diff
@@ -0,0 +1,462 @@
+Authentication and access delegations
+=====================================
+
+Introduction
+------------
+
+In this document, we will cover how to protect the usage of the cozy-stack.
+When the cozy-stack receives a request, it checks that the request is
+authorized, and if yes, it processes it and answers it.
+
+
+What about OAuth2?
+------------------
+
+OAuth2 is about delegating an access to resources on a server to another
+party. It is a framework, not a strictly defined protocol, for organizing the
+interactions between these 4 actors:
+
+- the resource owner, the "user" that can click on buttons
+- the client, the website or application that would like to access the
+  resources
+- the authorization server, whose role is limited to give tokens but is
+  central in OAuth2 interactions
+- the resources server, the server that controls the resources.
+
+For cozy, both the authorization server and the resources server roles are
+played by the cozy-stack. The resource owner is the owner of a cozy instance.
+The client can be the cozy-desktop app, cozy-mobile, or many other
+applications.
+
+OAuth2, and its extensions, is a large world. At its core, there is 2 things:
+letting the client get a token issued by the authorization server, and using
+this token to access to the resources. OAuth2 describe 4 flows, called grant
+types, for the first part:
+
+- Authorization code
+- Implicit grant type
+- Client credentials grant type
+- Resource owner credentials grant type.
+
+On cozy, only the most typical one is used: authorization code. To start this
+flow, the client must have a `client_id` and `client_secret`. The Cozy stack
+implements the OAuth2 Dynamic Client Registration Protocol (an extension to
+OAuth2) to allow the clients to obtain them.
+
+OAuth2 has also 3 ways to use a token:
+
+- in the query-string (even if the spec does not recommended it)
+- in the POST body
+- in the HTTP Authorization header.
+
+On cozy, only the HTTP header is supported.
```

Comment:
We will need to support query-string token at least for vfs download route to allow using vfs files in html `<audio>` tags. (see https://github.com/cozy/cozy-data-system/commit/f1f950eedba31699540de4ee99fc0fb65951a76d).

If we really dont want to, we will need a "create temporary unauthenticated route for this file" mechanism.